### PR TITLE
Partially applied type families

### PR DESCRIPTION
**Pull request by mniip**

### CODE REVIEW

**Comment by goldfirere** [comment]
[Rendered](https://github.com/mniip/ghc-proposals/blob/partiallyappliedtypefamilies/proposals/0000-partially-applied-type-families.rst)

**Comment by AntC2** [comment]
Thanks mniip, Richard. I'm commenting because nobody else is. Perhaps because ...

I'm finding your 'Motivation' entirely, er, unmotivating. 

>  One would be able to write type-level composition, folds, maps, zips, etc.

OK. Why would you want to do that? What would it help you write that you can't write currently? (Or that you could write but only awkwardly?) Delightful to read as Richard's Dissertation is, I didn't get any strong motivation from Section 4.2.4.

Is this some sort of intellectual ideal that type-level programming should be the same as term-level programming? I just don't see it.

More particularly: doing anything inside Haskell's type system is hard and usually breaks things. (When Type families were first introduced, Haskell took several steps backwards; and many releases to build it all out.) Why should this proposal get any attention over, say, impredicativity or the records system (which probably needs some type-level stuff, possibly impredicativity)? The lack of those is causing real pain; whereas I'm not hearing any anguish from people unable to write type-level folds.

**Comment by mniip** [comment]
> Is this some sort of intellectual ideal that type-level programming should be the same as term-level programming? I just don't see it.

Kind of, yes, wouldn't that be great? That's what dependent haskell is all about.

> I didn't get any strong motivation from Section 4.2.4.

The section was meant to be more of a theoretical basis than a motivation point. For motivation behind dependent types _and type-level programming in general_, see [Chapter 3](http://cs.brynmawr.edu/~rae/papers/2016/thesis/eisenberg-thesis.pdf#page=31).

> What would it help you write that you can't write currently?

Doing complicated stuff on the type-level comes up constantly. If you want a concrete example of why such advanced type-level computations are needed, recently I've sketched an applicative LR parser idea. The central datatype `Parser a`  contains a list of shift and a list of reduce rules, the problem is, reduce rules are functions of varying types, and, obviously you want to keep type safety, so you have to arrange this zoo of functions into a homogeneous list. You could use Typeable but then polymorphism goes out the window. My approach was to use a type-level tree-set of types. Paths in this set would designate a certain nonterminal and the set would contain the nonterminal's type at this path. In my setup, all reduce rules were of type `Reduce (m :: TMap *)`, and Reduce had a GADT constructor which contained the indices of the three involved nonterminals (two inputs and one output) as well as equality constraint proofs that the types at those indices in the map were indeed compatible.

I don't believe the above example is very contrived, so I think it is valid to say that type-level programming is an important part of haskell that we should pay attention to.

Now, the type map obviously had to have some non-trivial operations defined on it, etc. And so we're back to square one, functional programming, but now at type level. All core principles of functional programming still apply. Among them is generalization of algorithms using higher order functions.

> Why should this proposal get any attention over <...>

Proposals aren't a queue, and even if they are I'm certainly not "clogging" it and getting in the way of other (possibly more important) improvements.

**Comment by mniip** [comment]
> When Type families were first introduced, Haskell took several steps backwards; and many releases to build it all out.

Though, I should say that I agree that type families are kind of awkward. In contrast with dependent types they look very "half-baked".

**Comment by cartazio** [comment]
To my knowledge, type families allow type case like computations, which
aren't accessible in the same way with dependent types.

**Comment by mniip** [comment]
Dependent types let you lift arbitrary computations into the type level. Including case-matches.

**Comment by andrewthad** [comment]
@mniip Typecase refers to something different from ordinary pattern matching. Here's a [SO question with some answers that explain it](http://stackoverflow.com/questions/23220884/why-is-typecase-a-bad-thing).

That aside, I don't particularly want type families extended further in haskell. I would rather have dependent types. Just my preference.

**Comment by goldfirere** [comment]
More comments on the proposal in due course, but just to respond to @andrewthad : Designing and implementing this is on the direct path toward proper dependent types. After all, once we have term-level functions in types, those, too, will have to be able to appear unsaturated. I think very little of this effort would be made redundant by dependent types.

**Comment by AntC2** [comment]
>> When Type families were first introduced, Haskell took several steps backwards; and many releases to build it all out.

I should say that now Type families are largely complete, they're an impressive feature, with more power than I at first realised (or than is really documented).

@mniip 
> Though, I should say that I agree that type families are kind of awkward. In contrast with dependent types they look very "half-baked".

If there's dissatisfaction with type families, it'd be better to fix that before burdening them with more 'features' like partial application.

(I suspect my dissatisfaction with them will be on different grounds compared to yours.)

**Comment by goldfirere** [comment]
I like the overall direction of this proposal, with one quibble and one major shortcoming.

The quibble: I would like `~>` (or some other new name) to be the **matchable** arrow, and use `->` to be the **unmatchable** arrow. This is the reverse of the feature as proposed. My motivation is all to do with Dependent Haskell: when we write the type of an ordinary term-level function (e.g., `id :: a -> a`) we use `->`. But term-level functions are **unmatchable**. Under Dependent Haskell (a preliminary description of which is Chapter 4 of my [thesis](http://cs.brynmawr.edu/~rae/papers/2016/thesis/eisenberg-thesis.pdf)) terms and types will be indistinguishable, and so if we use different arrows to describe terms as we do types, there will be a challenge.

A drawback of my proposed change here is that it's not backward compatible. Right now, every arrow we use in a kind describing a type is matchable. (Note that we *never* write out the kind of a type family in today's Haskell. GHCi can do this in its `:kind` and `:info` operations, but these are special cases baked into `:kind` and `:info`.) So my proposed change here means that we would have to have some treatment of backward-compatibility, saying that when the user writes, e.g., `data T :: a -> b -> c -> Type`, those arrows are actually `~>`s, not `->`s. (The same would be true in data constructor types.) Perhaps we could require that the user write `~>` in these places with `-XCurriedTypeFamilies` enabled, as those modules need not be backward-compatible.

The major shortcoming: this proposal contains no specification of how kind inference should work. Furthermore, it says that the matchable arrow will be a subtype of the unmatchable arrow. While I agree with this subtyping relationship superficially, subtyping and type inference are notoriously hard to do at the same time. (Indeed, a chief motivator behind levity polymorphism was to eliminate sub-typing in GHC.) Here are some concrete inference puzzles. What kinds would you ascribe to the type variables?

```
f :: forall a b c. a b -> a c -> a Int
g :: forall a b. Proxy (Map a b) -> ()   -- that's the promoted form of map :: (a -> b) -> [a] -> [b]
h :: forall a b. a b -> b Int -> ()
```

More problems follow on from these: depending on how kind inference works on these examples, some type variables would be inherently ambiguous (that is, impossible to discover from a call site without `TypeApplications`). Will this be easily predictable by (sufficiently advanced) users? And does the sub-typing relation work contravariantly to the left of arrows (as do most sub-typing relationships)? There may be answers buried in Chapter 6 of my thesis, but it would take some work to simplify the issues brought up there for intelligent discussion here.

My bottom line: Thanks much for posting this and starting this conversation. Modulo the quibble, I agree with what you've written. But much more needs to be said about this before we have an implementable feature.

On incremental improvements / the "half-baked" nature of type families: I agree completely. Indeed, as I was teaching my undergrads this semester about fancy types in Haskell, I got many questions motivated by the weird place of type families in the language. It is my hope that Dependent Haskell will obviate a lot of these concerns, allowing us to deprecate type families as we know them (you would just use an ordinary function in types instead). Regardless, the challenges in implementing curried type families would persist in Dependent Haskell, with an apparent need for two arrows, a potential sub-typing relationship, and a specification for type inference. So, while this work, at its surface, is patching up a feature that seems in need of replacement, the replacement will benefit greatly from the work done here.

**Comment by mniip** [comment]
*Throughout this post I'm sticking with my notation of matchable `->` and unmatchable `~>`*

> The major shortcoming: this proposal contains no specification of how kind inference should work. Furthermore, it says that the matchable arrow will be a subtype of the unmatchable arrow. While I agree with this subtyping relationship superficially, subtyping and type inference are notoriously hard to do at the same time. (Indeed, a chief motivator behind levity polymorphism was to eliminate sub-typing in GHC.)

I was thinking of having something exactly like levity polymorphism. Something akin to:
```
data ARROW :: forall (r :: RuntimeRep) (s :: RuntimeRep). () -> TYPE r -> TYPE s -> Type
type (->) = forall (z :: ()). ARROW z
type (~>) = ARROW '()
```
I'm not sure exactly how to formalize this (because of the impredicative polymorphism). But basically I want to have:
```
map :: forall (a :: Type) (b :: Type). ARROW '() (ARROW '() a b) (ARROW '() a b)
IdentityT :: forall (y :: ()) (z :: ()). ARROW y (forall (x :: ()). ARROW x Type Type) (ARROW z Type Type)
```
Notice how IdentityT is Rank-2 -- it expects a matchable type function.

From right here I see a slight issue with typeclasses: `instance Functor ((->) e)` which is implicitly translated into `instance Functor ((~>) e)` requires FlexibleInstances.

As for your type puzzles, considering currently nowhere in the language we an quantify over a type function, I think a sensible default is to infer matchable kind in the situation like:
```
f x :: k, x :: l
----------------
  f :: l -> k
```

Then you would have:
```
f :: forall (a :: Type -> Type) (b :: Type) (c :: Type). a b ~> a c ~> a Int

Map :: forall (a :: Type) (b :: Type). (a ~> b) ~> [a] ~> [b]
-- therefore
g :: forall (a :: Type ~> Type) (b :: Type). Proxy (Map a b) ~> ()

h :: forall (a :: (Type -> Type) -> Type) (b :: Type -> Type). a b ~> b Int ~> ()
```

**Comment by mniip** [comment]
@goldfirere I just realized I did a bit of a mistake in `g`, see edited version.

**Comment by mniip** [comment]
You can actually kind of achieve/emulate this in modern GHC:
```
{-# LANGUAGE TypeInType, RankNTypes, TypeOperators, TypeFamilies, DataKinds #-}
{-# LANGUAGE UndecidableInstances #-}

import Data.Kind
import GHC.Exts

data ARROW :: forall (r :: RuntimeRep) (s :: RuntimeRep). () -> TYPE r -> TYPE s -> Type
-- cannot use (->) because that's built-in syntax
type (-->) a b = forall (z :: ()). ARROW z a b
type (~~>) = ARROW '()

-- juxtaposition operator
type family ($$) :: (k ~~> l) -> k -> l where

-- This is supposed to be 'data Proxy' but alas (-->) means nothing to the typechecker
type family Proxy :: (k --> Type) where

type family Map :: (a ~~> b) ~~> ([a] ~~> [b]) where

g :: Proxy $$ ((Map $$ a) $$ b) -> ()
g = g
```

This correctly infers
```
g :: forall a b (a1 :: a ~~> b) (b1 :: [a]). Proxy $$ ((Map $$ a1) $$ b1) -> ()
```

The only difference is that GHC does:
```
f $$ x :: k, x :: l
-------------------
    f :: l ~~> k
```
while I'm proposing a stronger statement:
```

f $$ x :: k, x :: l
-------------------
    f :: l --> k
```

However I'm not too sure if what I'm suggesting here is actually valid. Basically something tells me the inference rules will depend on whether you're looking in or out and I don't think that's good.

**Comment by AntC2** [comment]
Yes, @AntC2's reference is discussing/questioning the ``~>`` syntax https://github.com/ghc-proposals/ghc-proposals/pull/54#issuecomment-308919543

**Comment by AntC2** [comment]
Thanks @mniip for the explanation (on #54) of the twiddly arrow. I'd rather discuss that topic there, because this proposal seems to be confused about which arrow means what. Back to partially applied ...

Could we not say:
* In principle any type-making thingy might be partially applied.
* Then we need only one style of arrow.

* In practice, that might make kind-inference undecidable (esp for polymorphic recursion).
* So you must supply kind -- er, I mean -- type-type-level signatures, per #54.
* That's similar to term-level where if you use advanced features you need a signature.

* Also in practice certain type-makers can't be partially applied.
* So if they appear unsaturated, kind inference gets 'stuck' -- as happens at term level.
* So the programmer must supply more arguments, or perhaps a signature.

**Comment by nomeata** [comment]
There wasn’t any discussion here for a while. What is the status of this proposal?

File: proposals/0000-partially-applied-type-families.rst [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+  fun2 :: a -> a
+  fun2 x = x
+  -- fun :: a ~> a
+
+Pattern matching of function-kinded entities in type family RHS'es and typeclass instance bodies can only happen at `(->)` kind.
+
+.. code-block:: haskell
+
+  type family Id :: (a :: *) :: * where Id a = a
```
Comment (mizunashi-mana) [inline proposals/0000-partially-applied-type-families.rst]: @mniip  There is a redundant `::`

**Comment by andrewthad** [comment]
I agree with Richard that it would be better to use `~>` as the matchable arrow. Inference seems like the main thing the proposal failed to sort out, and I'm wondering if it's possible to kick that can down the road. Right now, GHC treats any type-level arrow as matchable. Would it be possible to default all type-level arrows to being matchable unless an annotation was explicitly provided?

**Comment by goldfirere** [comment]
I'm worried about not having a better story around inference. Predictable type inference is important to GHC. But perhaps it's easy here.... I don't have a counterexample.

**Comment by nomeata** [comment]
Discussion here seemed to have ebbed down. I am marking this as dormant, until discussion either picks up again, or someone submits this for review.

**Comment by nomeata** [comment]
@mniip nothing new is happening here. Do you want to abandon this? In this case, just close the PR.

**Comment by nomeata** [comment]
Closing due to inactivity, but it can of course be reactivated any time.

**Comment by andrewthad** [comment]
I was using backpack the other day and realized that this proposal would have a very pleasant interaction with backpack's module signatures. Currently, all data declarations (or more accurately type declarations) in a module signature are treated as being matchable. This is, when can write:

```
signature Foo where
  data Universe :: Type
  data Interpret :: Universe -> Type
```

But we cannot implement `Interpret` as a type family. With this proposal, we can write:

```
signature Foo where
  data Universe :: Type
  data Interpret :: Universe ~> Type
```

And then we could actually get what we want.

**Comment by simonpj** [comment]
There's a paper about this now: [Higher-order type-level programming in Haskell](https://www.microsoft.com/en-us/research/publication/higher-order-type-level-programming-in-haskell/)

**Comment by andrewthad** [comment]
This is great news! And the inference-related concerns are resolved in the paper! I had forgotten until now that at ICFP, Richard had mentioned that Csongor had shown him a branch for this that worked really well. Searching for "matchable" and "unsaturated" on the GHC gitlab, I cannot find an MR for this. Do you have any knowledge about the status of the implementation that the paper references?

cc @kcsongor

**Comment by jvanbruegge** [comment]
@simonpj Is there a typo in the paper?
> Definition (Generativity). f and g are generative <=> f a ∼ g b => f ∼ g.

Shouldn't it be `f a ~ g a => f ~ g` or am I understanding this wrong?

**Comment by Ericson2314** [comment]
@jvanbruegge not a typo. See the end for the weaker notion of generativity you expected.

Is the date a typo though? It says January 2018.

**Comment by andrewthad** [comment]
Any information on where I could find the branch implementing this feature?  I’m working on a library  where it would be interesting to get to try this out.

**Comment by andrewthad** [comment]
Sorry to poking, but does anyone happen to have a copy of the branch where this feature is implemented?

**Comment by kcsongor** [comment]
@andrewthad sure! I just uploaded my prototype implementation here: https://github.com/kcsongor/ghc/tree/unsaturated_type_families

There are a couple of wrinkles left, notably `Typeable` evidence generation does not quite work yet.

**Comment by andrewthad** [comment]
I took a few stabs at building the branch today. No matter what I did (freezing stage 1 to be a build of master, etc.), I couldn't get the stage 2 build to complete. It always gets hung up on the typeable stuff. Are there any serious issues that prevent a typeable instance for:

(->>) :: Matchability -> forall (rep1 :: RuntimeRep) (rep2 :: RuntimeRep). TYPE rep1 -> TYPE rep2 -> Type

I'd be happy to take a crack at it if you already know that there aren't any particularly bad dragons there.

**Comment by kcsongor** [comment]
That's my bad, something went wrong when I rebased the branch it seems. Anyhow, the latest should now bootstrap fine.

The main difficulty is that the kind of `->>` is, while a rank-1 kind, not in prenex normal form, which means that the runtime reps don't get instantiated to skolems during instance selection, because the `Matchability` argument gets in the way.

That said, I think it should just work if the kind instead was

```
(->>) :: forall (m :: Matchability) (rep1 :: RuntimeRep) (rep2 :: RuntimeRep). TYPE rep1 -> TYPE rep2 -> Type
```

but I didn't do that because when I first implemented this feature, GHC didn't have support for visible kind applications. But now that I've rebased it to a patch that does, it should be fairly simple to pull that off.

Either way, I think I'm going to open a fresh proposal instead of trying to revive this one.

**Comment by andrewthad** [comment]
Cool. Thanks for posting the newest code. After looking at the old code, I was going to mention that `TrApp` needs to be applications involving the matchable arrow only and `TrFun` should be matchability-polymorphic, but from the new code, it looks like you've already realized that. Thanks again. I'm really looking forward to using this.

**Comment by Ericson2314** [comment]
https://github.com/ghc-proposals/ghc-proposals/pull/242 for reference that paper begat this proposal.

**Pull request closed without merging**